### PR TITLE
unilang: init at 0.13

### PR DESCRIPTION
**Pull request by rc-zb**
## Description of changes

Add the reference implementation of [Unilang](https://github.com/linuxdeepin/unilang/), a general purpose programming language by [UnionTech Software Technology Co., Ltd. (统信软件技术有限公司)](https://www.uniontech.com/).

And add myself to the maintainer lists of Nixpkgs and this package.

### Limitations

- LLVM-based JIT is disabled for lack of LLVM 7 (the version specified by the upstream; I haven't tested whether some other versions work).
- The upstream [`init.txt`](https://github.com/linuxdeepin/unilang/blob/9c4eb374fcdd7af02f35f4730bb7c0db36fa836e/init.txt) that "implements the new high-level language syntaxes" is not included, for that it seems to be expected to be supplied by the user and [put in the current working directory](https://github.com/linuxdeepin/unilang/blob/9c4eb374fcdd7af02f35f4730bb7c0db36fa836e/doc/Interpreter.zh-CN.md#%E7%94%A8%E6%88%B7%E7%8E%AF%E5%A2%83%E5%88%9D%E5%A7%8B%E5%8C%96).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/un/unilang/package.nix [hunk 1]
```diff
@@ -0,0 +1,90 @@
+{ lib, stdenv, fetchFromGitHub
+, libffi
+, libsForQt5
+, lld
+, parallel
+, pkg-config
+}:
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:7]:
```suggestion
{ lib
, stdenv
, fetchFromGitHub
, libffi
, libsForQt5
, lld
, parallel
, pkg-config
}:
```

File: pkgs/by-name/un/unilang/package.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "unilang";
+  version = "0.13";
+  src = fetchFromGitHub {
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:12]:
```suggestion

  src = fetchFromGitHub {
```

File: pkgs/by-name/un/unilang/package.nix [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    pkg-config
+  ];
+
+  /*
+    `std.txt` is a component of Unilang installation, and must be kept in
+    the output.
+  */
+  preBuild = ''
+    mkdir -p $out/lib/
+    cp ./std.txt $out/lib/
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:35]: This two lines should be at installPhase

File: pkgs/by-name/un/unilang/package.nix [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    substituteInPlace ./src/Main.cpp \
+        --replace-fail \"std.txt\" \"$out/lib/std.txt\" \
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:37]: This should be at patchPhase

File: pkgs/by-name/un/unilang/package.nix [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+  '';
+  # LLVM-based JIT is disabled for lack of LLVM 7.
+  buildPhase = ''
+    runHook preBuild
+
+    export Unilang_Output=$out/bin/unilang
+    export Unilang_BuildDir=$(pwd)/build/
+    export UNILANG_NO_LLVM=0;
+    export LDFLAGS=-fuse-ld=lld
+    bash ./build.sh
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:47]:
```suggestion
    chmod +x ./build.sh
    ./build.sh
```

File: pkgs/by-name/un/unilang/package.nix [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    `std.txt` is a component of Unilang installation, and must be kept in
+    the output.
+  */
+  preBuild = ''
+    mkdir -p $out/lib/
+    cp ./std.txt $out/lib/
+    substituteInPlace ./src/Main.cpp \
+        --replace-fail \"std.txt\" \"$out/lib/std.txt\" \
+  '';
+  # LLVM-based JIT is disabled for lack of LLVM 7.
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:39]: Needs more spaces.

```suggestion
  '';

  # LLVM-based JIT is disabled for lack of LLVM 7.
```

File: pkgs/by-name/un/unilang/package.nix [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+      libsForQt5.qt5.qtbase
+      libsForQt5.qt5.qtdeclarative
+      stdenv.cc.cc.lib
+    ];
+  in [
+    "--suffix LD_LIBRARY_PATH : ${runtimeLibPath}"
+  ];
+
+  doCheck = true;
+  preCheck = ''
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:63]: Again, postPatch

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/un/unilang/package.nix [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  src = fetchFromGitHub {
+    owner = "linuxdeepin";
+    repo = "unilang";
+    rev = "V" + finalAttrs.version;
+    hash = "sha256-GWY9FtbX3+FrxRSKWPcJKXm43tWF6AHggjaLr3YtN/8=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
+    libsForQt5.qt5.qtbase
```
Comment (AndersonTorres) [inline pkgs/by-name/un/unilang/package.nix:24]:
```suggestion
    qt5.qtbase
```

It should work.

**Comment by wineee** [comment]
It's a pity that unilang has not been maintained for a year,  this project may be dead :(

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Squash the commits to remove the fixups, according to contributing guidelines

**Pull request closed without merging**